[16.0][ADD] - sale_order_line_cancel

sbejaoui opened the pull request:
This module allows you to cancel the remaining quantity on sale order by adding
a dedicated action to sale lines. It also add two new fields to track canceled
and remaining to deliver quantities.

sbejaoui committed:
[ADD] - sale_order_line_cancel

sbejaoui commented:
cc/ @lmignon

lmignon reviewed (changes requested):
@sbejaoui I tested the odoo native behavior and it seems that this addon could be simplified and only provide the functionality in a specific context. On a sale order line, you can edit the quantity even if the SO is validated... IMO this behavior is dangerous and should be forbidden. Hopefully, an option is available into the settings to prevent the modification of quantities when the sale order is confirmed... It's only in the context where the modification is no more possible that this functionality makes sens and should be activated. Moreover, I've the feeling that all the custom code written to cancel the moves could be removed (and the dependency on https://github.com/OCA/stock-logistics-workflow/pull/1202). When you change the qty on a confirmed line, Odoo takes the necessary steps to reflect this change in the picking chain. The only thing Odoo doesn't do is to prevent the cancel of quantities that are partially processed. Instead, it generate the return picking to revert this outgoing moves... I prefer to prevent the cancellation of qties already partially processed than adding more work to do by opertors. This behaviour could be a configurable option..

sbejaoui commented:
> @sbejaoui I tested the odoo native behavior and it seems that this addon ..

Thanks @lmignon for your input,

I understand your PoV, It's true that the change of the ordered quantity trigger the change/cancellation of the related moves.

This feature is meant to simplify the process and make it done with a simple click. It also keep the information of the initial ordered quantity and store the canceled quantity. The difference between the two fields may be helpful for sale reporting.

The dependency to https://github.com/OCA/stock-logistics-workflow/pull/1202 is not a must. It's, IMO, a simpler way to get all sale line related moves instead of recursive loop.

lmignon commented:
> > @sbejaoui I tested the odoo native behavior and it seems that this addon ..
> 
> Thanks @lmignon for your input,
> 
> I understand your PoV, It's true that the change of the ordered quantity trigger the change/cancellation of the related moves.
> 
> This feature is meant to simplify the process and make it done with a simple click. It also keep the information of the initial ordered quantity and store the canceled quantity. The difference between the two fields may be helpful for sale reporting.
> 
> The dependency to [OCA/stock-logistics-workflow#1202](https://github.com/OCA/stock-logistics-workflow/pull/1202) is not a must. It's, IMO, a simpler way to get all sale line related moves instead of recursive loop.

IMO, the functionality should only be available if the change of qty on confirmed sale order is activated..

lmignon commented on sale_cancel_remaining/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,3 @@
+This module allows you to cancel the remaining quantity on sale order by adding
+a dedicated action to sale lines. It also add two new fields to track canceled
+and remaining to deliver quantities.
```
You should add:

```
This module differs from the original odoo behavior in the following way:

* In odoo, if the update of the qty ordered is allowed on the sale order even 
  when the order is confirmed, odoo will recompute the required stock operations 
  according to the changes done to the qty ordered. This change is possible
  even the stock operations are started for this sale order line.
* In this module, the qty ordered is not updated on the sale order line to keep
  the original qty ordered. At the same time, we cancel only the stock moves
  for the remaining qty to deliver. This is only possible if no operation is
  started for this sale order line.


.. warning::
    
    It's not recommended to use this module at the same time that allowing
    the update of the qty ordered on the sale order line after the order is
    confirmed. This could lead to unpredictable behavior.

```

lmignon commented on sale_cancel_remaining/wizards/__init__.py:
```diff
@@ -0,0 +1,3 @@
+# Copyright 2018 Okia SPRL
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
```
copyright is useless here

lmignon reviewed (approved):
Some comment to add to the desciption

jbaudoux commented on sale_cancel_remaining/views/sale_order_line.xml:
```diff
@@ -0,0 +1,21 @@
+    <field name="inherit_id" ref="sale.view_order_line_tree" />
+    <field name="arch" type="xml">
+      <xpath expr="//field[@name='product_uom_qty']" position="after">
+        <field name="product_qty_canceled" />
```
As the other fields also have a string to display Order | Delivered | Invoiced
```suggestion
        <field name="product_qty_canceled" string="Canceled"/>
```

jbaudoux commented on sale_cancel_remaining/views/sale_order_line.xml:
```diff
@@ -0,0 +1,21 @@
+    <field name="arch" type="xml">
+      <xpath expr="//field[@name='product_uom_qty']" position="after">
+        <field name="product_qty_canceled" />
+        <field name="product_qty_remains_to_deliver" />
```
```suggestion
        <field name="product_qty_remains_to_deliver" string="To Deliver"/>
```

jbaudoux commented on sale_cancel_remaining/wizards/cancel_remaining_wizard.py:
```diff
@@ -0,0 +1,56 @@
+
+    @api.model
+    def _get_moves_to_cancel(self, line):
+        moves_to_cancel = line.mapped("move_ids").filtered(
```
```suggestion
        moves_to_cancel = line.move_ids.filtered(
```

jbaudoux commented on sale_cancel_remaining/wizards/cancel_remaining_wizard.py:
```diff
@@ -0,0 +1,56 @@
+
+
+class CancelRemainingWizard(models.TransientModel):
+    _name = "cancel.remaining.wizard"
```
Maybe safer to bind the wizard to the model it applies to
```suggestion
    _name = "sale.order.line.cancel"
```

jbaudoux reviewed (commented):
I did many tests and it works quite well. It restricts the possibility of changing the quantity on the sale order line but at least the cancelation is working properly while in standard odoo you can end-up with a very weird reverse picking as decreasing the sale order line quantity generates a negative move that is supposed to be merged with the original move but if it can't be assigned to the same picking, it will create a reverse picking.

A few remarks that would make this module more complete:
* when you set the sales order back to quotation, `product_qty_canceled` on the lines should be reset to 0
* I would display the cancel button on a locked sales order as you expect auto-locking to be enabled and you allow canceling the lines
* When canceling the sales order, I would also write `product_qty_remains_to_deliver` in `product_qty_canceled`
* If you don't generate the backorder of the outgoing picking, the `product_qty_canceled` on the lines should also increase

jbaudoux commented:
> The dependency to [OCA/stock-logistics-workflow#1202](https://github.com/OCA/stock-logistics-workflow/pull/1202) is not a must. It's, IMO, a simpler way to get all sale line related moves instead of recursive loop.

Just a side note that this way of finding the right moves of the chain is not safe as moves can be merged. For example, a single pick move could serve multiple delivery moves. This specific use case is covered by the stock_available_to_promise_release module which, in case you cancel the outgoing move, will unrelease the outgoing move and while unreleasing it, will go through the picking chain up to the source picking and split any merged move.

A solution would be that this module only cancel the outgoing move and have another module using the dependency to [OCA/stock-logistics-workflow#1202](https://github.com/OCA/stock-logistics-workflow/pull/1202) to propagate the cancelation.

You can maybe add this remark in a ROADMAP file.

jbaudoux commented on sale_cancel_remaining/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{
+    "name": "Sale Cancel Remaining",
```
I would rename this module to sale_order_line_cancel

jbaudoux reviewed (approved):
Thanks for the improvements
LGTM

rousseldenis commented on sale_order_line_cancel/models/sale_order_line.py:
```diff
@@ -0,0 +1,43 @@
+    def _compute_can_cancel_remaining_qty(self):
+        for rec in self:
+            rec.can_cancel_remaining_qty = (
+                rec.product_qty_remains_to_deliver > 0
```
You should use float_compare

sbejaoui replied:
done

rousseldenis reviewed (changes requested):
Code review

jbaudoux commented on sale_order_line_cancel/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+{
+    "name": "Sale Order Line Cancel",
+    "version": "16.0.1.0.0",
+    "author": "Camptocamp, ACSONE SA/NV, Odoo Community Association (OCA)",
```
Let's keep original authors
```suggestion
    "author": "Okia, BCIM, Camptocamp, ACSONE SA/NV, Odoo Community Association (OCA)",
```

cyrilmanuel commented:
ping @sbejaoui :)

sbejaoui committed:
[ADD] sale_order_line_cancel: add authors

Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

sbejaoui committed:
[IMP] sale_order_line_cancel: use float_compare

cyrilmanuel commented:
thank you @sbejaoui  :+1:  @rousseldenis  can yo check the PR ?

sbejaoui committed:
[IMP] trace sale line cancel

sbejaoui committed:
[IMP] sale_order_line_cancel: check moves before cancel

lmignon committed:
[IMP] sale_order_line_cancel: Improve cancel process

Cancel remaining qty by running negative procurement orders to let odoo generate the appropriate operations

jbaudoux commented on sale_order_line_cancel/README.rst:
```diff
@@ -0,0 +1,96 @@
+
+* Sylvain Van Hoof <sylvain@okia.be>
+* Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
+* Souheil Bejaoui <souheil.bejaoui@acsone.eu.com>
```
```suggestion
* Souheil Bejaoui <souheil.bejaoui@acsone.eu>
```

sbejaoui committed:
[FIX] readme

Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

jbaudoux commented on sale_order_line_cancel/models/stock_move.py:
```diff
@@ -0,0 +1,20 @@
+        for rec in sale_moves:
+            if rec.state != "cancel":
+                continue
+            rec.sale_line_id.product_qty_canceled = rec.product_uom_qty
```
This is not correct for the kits (mrp)

jbaudoux replied:
we should also ensure on sale.order.line that `product_qty_canceled <= product_uom_qty - qty_delivered`

jbaudoux commented on sale_order_line_cancel/models/sale_order.py:
```diff
@@ -0,0 +1,34 @@
+
+    def _action_cancel(self):
+        orders = self.filtered(lambda s: s.state != "cancel")
+        orders_with_picking = orders.filtered("picking_ids")
```
Maybe this is better as in `sale_stock` they use `qty_delivered_method`:
```suggestion
        orders_with_picking = orders.filtered(lambda o: any(line.qty_delivered_method == 'stock_move' for line in o.order_line)):
```

jbaudoux reviewed (commented):
We should also support dropship & kits
in regards of https://github.com/odoo/odoo/blob/16.0/addons/sale_mrp/models/sale_order_line.py#L33

cyrilmanuel commented:
Hi @sbejaoui little ping, could you check about the comment of @jbaudoux  ?

jbaudoux committed:
[IMP] sale_order_line_cancel: support kits

jbaudoux commented:
See https://github.com/acsone/sale-workflow/pull/4

rousseldenis committed:
Revert "[IMP] sale_order_line_cancel: Improve cancel process"

This reverts commit b5cd7f03b1ac6c5fe0c6e9ed582031709602443f.

rousseldenis committed:
[FIX] sale_order_line_cancel: Restore some features

Allows to check the cancellation on sale order line level

Include normal sales (not locked)

cyrilmanuel commented on sale_order_line_cancel/security/sale_order_line_cancel.xml:
```diff
@@ -0,0 +1,14 @@
+        <field name="perm_create" eval="1" />
+        <field name="perm_write" eval="1" />
+        <field name="perm_unlink" eval="1" />
+    </record>
```
is it possible to add  entry for users in groups  sales_team.group_sale_salesman ?

<record model="ir.model.access" id="sale_order_line_cancel_access_own_document">
        <field name="name">sale.order.line.cancel access own document</field>
        <field name="model_id" ref="model_sale_order_line_cancel" />
        <field name="group_id" ref="sales_team.group_sale_salesman" />
        <field name="perm_read" eval="1" />
        <field name="perm_create" eval="1" />
        <field name="perm_write" eval="1" />
        <field name="perm_unlink" eval="1" />
    </record>

rousseldenis commented on sale_order_line_cancel/wizards/sale_order_line_cancel.py:
```diff
@@ -0,0 +1,27 @@
+    _description = "Cancel Remaining Wizard"
+
+    def _get_sale_order_line(self):
+        active_id = self._context.get("active_id")
```
`_context` is deprecated

```suggestion
        active_id = self.env.context.get("active_id")
```

sbejaoui commented on sale_order_line_cancel/models/sale_order_line.py:
```diff
@@ -0,0 +1,77 @@
+        for line in self:
+            if line._get_moves_to_cancel():
+                continue
+            line.product_qty_canceled = line.qty_to_deliver
```
@jbaudoux , 

The canceled quantity is not the same as the quantity to deliver because the quantity to deliver includes returned quantities. 
For example, in an order of 10 units, if I deliver 8 and then cancel the remaining 2, and later my customer returns 2 units, I would expect the following values:

- Quantity to deliver: 4
- Quantity canceled: 2

lmignon committed:
[IMP] sale_order_line_cancel: Add hook method

The new hook method allows you to filter out move that should no be taken into account to trigger the update of the canceled qty on the sale order line

jbaudoux commented:
@rousseldenis Is this now also good for you? Can you update your review?

rousseldenis commented:
/ocabot merge nobump
